lib/Fhp/Response/Response.php

```diff
@@ -13,219 +13,214 @@
-    }
-
-    /**
-     * @return int
-     */
-    public function getHksalMaxVersion()
-    {
-        return $this->getSegmentMaxVersion('HISALS');
-    }
-
-    /**
-     * @return int
-     */
-    public function getHkkazMaxVersion()
-    {
-        return $this->getSegmentMaxVersion('HIKAZS');
-    }
-
-    /**
-     * Checks if request / response was successful.
-     *
-     * @return bool
-     */
-    public function isSuccess()
-    {
-        $summary = $this->getMessageSummary();
-
-        foreach ($summary as $code => $message) {
-            if ("9" == substr($code, 0, 1)) {
-                return false;
-            }
-        }
-		
-        $summary = $this->getSegmentSummary();
-
-        foreach ($summary as $code => $message) {
-            if ("9" == substr($code, 0, 1)) {
-                return false;
-            }
-        }
-
-        return true;
-    }
-
-    /**
-     * @return array
-     * @throws \Exception
-     */
-    public function getMessageSummary()
-    {
-        return $this->getSummaryBySegment('HIRMG');
-    }
-
-    /**
-     * @return array
-     * @throws \Exception
-     */
-    public function getSegmentSummary()
-    {
-        return $this->getSummaryBySegment('HIRMS');
-    }
-
-    /**
-     * @param string $name
-     *
-     * @return array
-     * @throws \Exception
-     */
+	/**
+	 * Extracts dialog ID from response.
+	 *
+	 * @return string|null
+	 * @throws \Exception
+	 */
+	public function getDialogId()
+	{
+		$segment = $this->findSegment('HNHBK');
+
+		if (null === $segment) {
+			throw new \Exception('Could not find element HNHBK. Invalid response?');
+		}
+
+		return $this->getSegmentIndex(4, $segment);
+	}
+
+	/**
+	 * Extracts bank name from response.
+	 *
+	 * @return string|null
+	 */
+	public function getBankName()
+	{
+		$bankName = null;
+		$segment = $this->findSegment('HIBPA');
+		if (null != $segment) {
+			$split = $this->splitSegment($segment);
+			if (isset($split[3])) {
+				$bankName = $split[3];
+			}
+		}
+
+		return $bankName;
+	}
+
+	/**
+	 * Some kind of HBCI pagination.
+	 *
+	 * @param AbstractMessage $message
+	 *
+	 * @return array
+	 */
+	public function getTouchDowns(AbstractMessage $message)
+	{
+		$touchdown = array();
+		$messageSegments = $message->getEncryptedSegments();
+		/** @var AbstractSegment $msgSeg */
+		foreach ($messageSegments as $msgSeg) {
+			$segment = $this->findSegmentForReference('HIRMS', $msgSeg);
+			if (null != $segment) {
+				$parts = $this->splitSegment($segment);
+				// remove header
+				array_shift($parts);
+				foreach ($parts as $p) {
+					$pSplit = $this->splitDeg($p);
+					if ($pSplit[0] == 3040) {
+						$td = $pSplit[3];
+						$touchdown[$msgSeg->getName()] = $td;
+					}
+				}
+			}
+		}
+
+		return $touchdown;
+	}
+
+	/**
+	 * Extracts supported TAN mechanisms from response.
+	 *
+	 * @return array
+	 */
+	public function getSupportedTanMechanisms()
+	{
+		$gv = new GetVariables($this->response);
```

Comment:
Instead of parsing the segments again this now uses the already available functionality in `GetVariables`. It also adds the mechanism names to the result which can be presented to the user in a GUI.

Reply:
:heavy_check_mark: